pkg/api/types.go

```diff
@@ -2128,3 +2128,35 @@ type RangeAllocation struct {
 	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
 	Data []byte `json:"data"`
 }
+
+// A ThirdPartyObject is a generic representation of a resource, it is used by add-ons and plugins to add new resource
+// types to the API.  It consists of one or more Versions of the api.
+type ThirdPartyObject struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty" description:"standard object metadata"`
+
+	Description string `json:"description,omitempty" description:"The description of this object"`
+
+	Versions []Version `json:"versions,omitempty" description:"The versions for this third party object"`
```

Comment:
* Key prefix added to design document
   * The version stored is the version that they give us (for now)
   * For now, we don't have any ability to convert between versions, ultimately that will be an HTTP web hook.  For now, it is on the user, and if you ask for a `v2` object and we stored a `v1` object, we will give you an error.  You can do the `v1` to `v2` conversion yourself manually using the same read/write pattern that we used for our own API upgrades.
   * For the reason for multiple versions, we need to know how to program multiple different HTTP paths (and store things in different formats, as above)

For Kind, I think we have to infer it from the name of the resource, if we don't we're going to be forced to either scan all Kinds or maintain a cache inside the API server to do the right thing when we receive a request for a custom object. 

Either of those things leads to a bunch of complexity in the implementation.

I think a better approach is to allow for domain names in the `Kind` specification for custom objects.  (e.g. `example.com/CamelCaseKind`) this will allow for the kinds of things that you want.  See the design doc for a complete discussion of this.